src/components/Timer.react.js

```diff
@@ -0,0 +1,322 @@
+ * Update messages clientside to help improve app performance.
+ * Specify the elapsed times to fire a callback rather than every interval
+ **/
+
+export default class Timer extends Component {
+    constructor(props) {
+        super(props);
+        this.intervalId = null;
+        this.renderMessage = null;
+        this.reportInterval = this.reportInterval.bind(this);
+        this.handleTimer = this.handleTimer.bind(this);
+        this.handleMessages = this.handleMessages.bind(this);
+
+        // this.countdown is used to handle the timer within this component.  If it's in stopwatch mode
+        //   then the count-up time is calculated just prior to setProps({timer:})
+        this.countdown = 0;
+    }
+
+    handleTimer(props) {
+        const {
+            n_intervals,
+            max_intervals,
+            disabled,
+            rerun,
+            interval,
+        } = this.props;
+
+        // Check if timer should stop or shouldn't even start
+        if (
+            max_intervals === 0 ||
+            disabled ||
+            (n_intervals >= max_intervals && max_intervals !== -1) ||
+            (this.countdown === 0 && !rerun) ||
+            (this.countdown > 0 && this.countdown < interval)
+        ) {
+            // stop existing timer
+            if (this.intervalId) {
+                this.clearTimer();
+            }
+            // and don't start a timer
+            return;
+        }
+
+        // keep the existing timer running
+        if (this.intervalId) {
+            if (this.countdown === 0 && rerun) {
+                this.initTimer(props);
+            }
+            return;
+        }
+
+        // it hasn't started yet (& it should start)
+        this.intervalId = window.setInterval(this.reportInterval, interval);
+    } // end handle timer
+
+    handleMessages(props, new_timer) {
+        const {messages, timer_format} = this.props;
+
+        const messagesObj = Object.assign({}, messages);
+        if (new_timer in messagesObj) {
+            this.renderMessage = `${messagesObj[new_timer]}`;
+        }
+
+        const prettyMilliseconds = require('pretty-ms');
+        if (timer_format !== 'none') {
+            const formatObj = getFormat(timer_format);
+            this.renderMessage = `${prettyMilliseconds(new_timer, formatObj)}`;
+        }
+    } // end handleMessages
+
+    reportInterval() {
+        const {
+            setProps,
+            n_intervals,
+            interval,
+            mode,
+            duration,
+            fire_times,
+        } = this.props;
+
+        const new_n_intervals = n_intervals + 1;
+        const updateProps = {n_intervals: new_n_intervals};
+
+        this.countdown = duration - interval * new_n_intervals;
+        let new_timer;
+        if (mode === 'countdown' && duration !== -1) {
+            new_timer = this.countdown;
+        } else {
+            // stopwatch.  Note - stopwatch mode only if duration === -1
```

Comment:
This logic seems confusing to me - also fragile, there's nothing stopping a user from setting some other negative duration. I think it would be cleaner if `countdown` mode treated any negative duration the same as 0, ie it's still in countdown mode but never starts.

Reply:
Right, that would work better.  I was trying to make it consistent with the `max_intervals`, which when set to -1 has no limit.  But it doesn't make as much sense for the `duration`.  Maybe if the `duration` is `None`, then the timer would work the same way as  `dcc.Interval`  (ie a count up timer with no limit on the duration)